src/sync/channel.zig

```diff
@@ -1,170 +1,426 @@
 const std = @import("std");
 const Atomic = std.atomic.Value;
-const Condition = std.Thread.Condition;
-const testing = std.testing;
+const Allocator = std.mem.Allocator;
 const assert = std.debug.assert;
 const Mux = @import("mux.zig").Mux;
+const Backoff = @import("backoff.zig").Backoff;
 
-/// A very basic mpmc channel implementation - TODO: replace with a legit channel impl
-pub fn Channel(comptime T: type) type {
+pub fn Channel(T: type) type {
     return struct {
-        buffer: Mux(std.ArrayList(T)),
-        has_value: Condition = .{},
+        head: Position,
+        tail: Position,
         closed: Atomic(bool) = Atomic(bool).init(false),
-        allocator: std.mem.Allocator,
+        allocator: Allocator,
 
         const Self = @This();
+        const BLOCK_CAP = 31;
+        const SHIFT = 1;
+        const LAP = 32;
 
-        pub fn init(allocator: std.mem.Allocator, init_capacity: usize) *Self {
-            const self = allocator.create(Self) catch unreachable;
-            self.* = .{
-                .buffer = Mux(std.ArrayList(T)).init(std.ArrayList(T).initCapacity(allocator, init_capacity) catch unreachable),
+        const WRITE: usize = 0b01;
+        const READ: usize = 0b10;
```

Comment:
Same as above. I might call this something like "FINISHED_READING"

Reply:
Renamed to `READ_FROM`.